src/GraphQL.SystemTextJson/ObjectDictionaryConverter.cs

```diff
@@ -0,0 +1,223 @@
+            JsonSerializerOptions options)
+        {
+            JsonSerializer.Serialize(writer, value, _internalOptions);
+        }
+
+        private Dictionary<string, object> ReadDictionary(JsonElement element)
+        {
+            var result = new Dictionary<string, object>();
+            foreach (var property in element.EnumerateObject())
+            {
+                var key = property.Name;
+                var value = property.Value;
+                object resultValue = null;
+
+                switch (value.ValueKind)
+                {
+                    case JsonValueKind.Object:
+                        resultValue = ReadDictionary(value);
+                        break;
+                    case JsonValueKind.Number:
+                        resultValue = ReadNumber(value);
+                        break;
+                    case JsonValueKind.True:
+                    case JsonValueKind.False:
+                        resultValue = value.GetBoolean();
+                        break;
+                    case JsonValueKind.String:
+                        resultValue = value.GetString();
+                        break;
+                    case JsonValueKind.Null:
+                        // default value is null
+                        break;
+                    case JsonValueKind.Array:
+                        resultValue = ReadArray(value).ToList();
+                        break;
+                    default:
+                        throw new InvalidOperationException($"Unexpected value kind: {value.ValueKind}");
+                }
+
+                result[key] = resultValue;
+            }
+
+            if (!result.Any())
+                return null;
+
+            return result;
+        }
+
+        private IEnumerable<object> ReadArray(JsonElement value)
+        {
+            foreach (JsonElement item in value.EnumerateArray())
+            {
+                switch (item.ValueKind)
+                {
+                    case JsonValueKind.Object:
+                        yield return ReadDictionary(item);
+                        break;
+                    case JsonValueKind.Number:
+                        yield return ReadNumber(item);
+                        break;
+                    case JsonValueKind.True:
+                    case JsonValueKind.False:
+                        yield return item.GetBoolean();
+                        break;
+                    case JsonValueKind.String:
+                        yield return item.GetString();
+                        break;
+                    case JsonValueKind.Null:
+                        yield return null;
+                        break;
+                    case JsonValueKind.Array:
+                        yield return ReadArray(item).ToList();
+                        break;
+                    default:
+                        throw new InvalidOperationException($"Unexpected value kind: {item.ValueKind}");
```

Comment:
Copy/paste from `ReadDictionary`. Consider to move in method like `ReadValue`

Reply:
Done - will push in a sec.